rangefeed: add memory adjuster before publishing events

wenyihu6 opened the pull request:
We over-accounts at the beginning to account for everything that could have happened. If the memory turns out to be not needed, we de-allocates the memory right away to prevent registrations from holding onto unnecessary memory for too long.

wenyihu6 committed:
rangefeed: add memory adjuster before publishing events

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120348)
<!-- Reviewable:end -->

wenyihu6 committed:
rangefeed: improve memory accounting for event queue

To prevent OOMs, we previously introduced memory budget to limit event buffering
to processor.eventC channel. These events could point to large underlying data
structure and keeping them in the channel prevent them from being garbage
collected. However, the current memory accounting did not have a good way to
predict the memory footprint of events. It uses calculateDateSize which 1. does
not account for memory of the new RangeFeedEvents that could be spawned (such as
checkpoint events) 2. inaccurately uses protobuf-encoded wire size for data 3.
does not account for memory of the base structs.

This patch improves the memory estimation by resolving the three points above.
It tries its best effort to account for base struct's memory, predict generation
of new events, and use actual data size rather than compressed protobuf size.
Since it is challenging to predict whether there would be a new checkpoint
event, our strategy in this PR is over-accounting >> under-accounting. GitHub
issue (todo) tracks the remaining work to improve accuracy.

Resolves: https://github.com/cockroachdb/cockroach/issues/114190
Release note: None
